org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/StandardPreferenceManager.java

```diff
@@ -198,9 +193,19 @@ public void update(Preferences preferences) {
 			updateMavenProjects = true;
 		}
 		if (updateMavenProjects) {
-			for (IProject project : ProjectUtils.getAllProjects()) {
-				if (ProjectUtils.isMavenProject(project)) {
-					JavaLanguageServerPlugin.getProjectsManager().updateProject(project, true);
+			ProjectsManager projectManager = JavaLanguageServerPlugin.getProjectsManager();
+			if (projectManager != null) {
+				if (projectManager.isBuildFinished()) {
+					projectManager.updateMavenProjects(null);
+				} else {
+					boolean hasMavenProjects = false;
+					for (IProject project : ProjectUtils.getAllProjects()) {
+						if (ProjectUtils.isMavenProject(project)) {
+							hasMavenProjects = true;
+							break;
+						}
+					}
+					projectManager.setShouldUpdateProjects(hasMavenProjects);
```

Comment:
Wouldn't the part in the `else` always need to run because [`update(Preferences)`](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/185a7cd03211518dfb1b0c91a0e816c48dd7ccc7/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseInitHandler.java#L115) is guaranteed to be called from `BaseInitHandler.handleInitializationOptions(..)` ? That runs before the build is finished so this is definitely always triggered.

If so then I would think you don't need the `shouldUpdateProjects` API, but let me know if I'm wrong.

Reply:
We need it. I want to delay updating the projects after the initialization of Java LS.
See 
- https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3150/commits/cfb076c10a86523d734942f29364d0d90aa6634f#diff-f82616576dc1fa363177feaadbbde3568ec33ec20588a8e7656ba45ded34bc52R332
- https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3150/commits/cfb076c10a86523d734942f29364d0d90aa6634f#diff-f23d0258ce24117c110acd6753e8490a09fbdf0bc039d662bdffe279c8c28469R671